File: lessons/ads/power_set.py
from typing import List, Optional


class PowerSet():
    def __init__(self) -> None:
        self.step = 1
        self.slots = {}

    def size(self) -> int:
        return len(self.slots)

    def hash_fun(self, value: str) -> int:
        # в качестве value поступают строки!
        # всегда возвращает корректный индекс слота
        return hash(value)

    def seek_slot(self, value: str) -> Optional[int]:
        # находит индекс пустого слота для значения, проверяя уникальность, или None
        slot = self.hash_fun(value)
        while slot in self.slots.keys():
            if self.slots[slot] == value:
                return None
            slot = slot + self.step
        return slot

    def put(self, value: str) -> Optional[int]:
        slot = self.seek_slot(value)
        if slot is None:
            return None
        self.slots[slot] = value
        return slot

    def find(self, value: str) -> Optional[int]:
        # находит индекс слота со значением, или None
        slot = self.hash_fun(value)
        while slot in self.slots.keys():
            if self.slots[slot] == value:
                return slot
            slot = slot + self.step
        return None

    def get(self, value: str) -> bool:
        # возвращает True если value имеется в множестве,
        # иначе False
        return self.find(value) is not None

    def remove(self, value: str) -> bool:
        # возвращает True если value удалено
        # иначе False
        slot = self.find(value)
        if slot is not None:
            self.slots.pop(slot)
            return True
        return False

    def intersection(self, set2: "PowerSet") -> "PowerSet":
        # пересечение текущего множества и set2
        intersection_set = PowerSet()
        for value in set2.slots.values():
            if self.get(value):
                intersection_set.put(value)
        return intersection_set

    def union(self, set2: "PowerSet") -> "PowerSet":
        # объединение текущего множества и set2
        union_set = PowerSet()
        for value in self.slots.values():
            union_set.put(value)
        for value in set2.slots.values():
            if not self.get(value):
                union_set.put(value)
        return union_set

    def difference(self, set2: "PowerSet") -> "PowerSet":
        # разница текущего множества и set2
        difference_set = PowerSet()
        if self.size() == 0:
            return difference_set
        for value in self.slots.values():
            difference_set.put(value)
        for value in set2.slots.values():
            difference_set.remove(value)
        return difference_set

    def issubset(self, set2: "PowerSet") -> bool:
        # возвращает True, если set2 есть
        # подмножество текущего множества,
        # иначе False
        if self.size() < set2.size():
            return False
        for value in set2.slots.values():
            if not self.get(value):
                return False
        return True

File: lessons/ads3_go/PowerSet/PowerSet.go
//10

package main

import (
	"constraints"
	"os"
	"strconv"
)

type IPowerSet[T any] interface {
	Size() int
	Remove(value T) bool
	Get(value T) bool
	Put(value T)
	Intersection(set2 IPowerSet[T]) IPowerSet[T]
	Union(set2 IPowerSet[T]) IPowerSet[T]
	Difference(set2 IPowerSet[T]) IPowerSet[T]
	IsSubset(set2 IPowerSet[T]) bool
	Equals(set2 IPowerSet[T]) bool
}

type PowerSet[T constraints.Ordered] struct {
	// ваша реализация хранилища
	// я искал в Go решение для использования нормальной хэш-функции и наткнулся на распространенный прием - мапа с пустой структурой в качестве значения.
	// Не совсем сет, подход такой: проверяем по ключу налчие значения в мапе. Мапа под капотом применяет встроенную хэш-функцию и на ее основе однозначно возвращает наличие или отсутствие значения в словаре.
	setSlots map[T]struct{}
	size     int
}

// создание экземпляра множества
func Init[T constraints.Ordered]() PowerSet[T] {
	return PowerSet[T]{setSlots: make(map[T]struct{}), size: 0}
}

func (p *PowerSet[T]) Size() int {
	// количество элементов в множестве
	return p.size
}

// О(1)
func (p *PowerSet[T]) Put(value T) {
	if _, exists := p.setSlots[value]; !exists {
		p.setSlots[value] = struct{}{}
		p.size++
	}
}

// О(1)
func (p *PowerSet[T]) Get(value T) bool {
	// возвращает true если value имеется в множестве
	_, exists := p.setSlots[value]
	return exists
}

// О(1)
func (p *PowerSet[T]) Remove(value T) bool {
	// возвращает true если value удалено
	if _, exists := p.setSlots[value]; exists {
		delete(p.setSlots, value)
		p.size--
		return true
	}
	return false
}

// О(n) на создание нового множества, O(n) по времени на итерацию по множеству
func (p *PowerSet[T]) Intersection(set2 PowerSet[T]) PowerSet[T] {
	// пересечение текущего множества и set2
	var result PowerSet[T]

	result = Init[T]()
	for key := range p.setSlots {
		if set2.Get(key) {
			result.Put(key)
		}
	}
	return result
}

// О(n+m) на создание нового множества, O(n+m) по времени на итерацию по множествам
func (p *PowerSet[T]) Union(set2 PowerSet[T]) PowerSet[T] {
	// объединение текущего множества и set2
	var result PowerSet[T]

	result = Init[T]()
	for key := range p.setSlots {
		result.Put(key)
	}
	for key := range set2.setSlots {
		result.Put(key)
	}
	return result
}

// О(n) на создание нового множества, O(n) по времени на итерацию по множеству
func (p *PowerSet[T]) Difference(set2 PowerSet[T]) PowerSet[T] {
	// разница текущего множества и set2
	var result PowerSet[T]

	result = Init[T]()
	for key := range p.setSlots {
		if !set2.Get(key) {
			result.Put(key)
		}
	}
	return result
}

// O(n) по времени на итерацию по множеству
func (p *PowerSet[T]) IsSubset(set2 PowerSet[T]) bool {
	// возвращает true, если set2 есть подмножество текущего множества
	if set2.Size() > p.Size() {
		return false
	}

	for key := range set2.setSlots {
		if !p.Get(key) {
			return false
		}
	}
	return true
}

// O(n) по времени на итерацию по множеству
func (p *PowerSet[T]) Equals(set2 PowerSet[T]) bool {
	// возвращает true, если set2 равно текущему множеству
	if p.Size() != set2.Size() {
		return false
	}
	for key := range p.setSlots {
		if !set2.Get(key) {
			return false
		}
	}
	return true
}
